README.md
Deep Learning Virtual Machine, `xenial64`
===============================

This is a VirtualBox VM that is automatically generated using Vagrant.

Machine Learning and Deep Learning packages installed are: Scikit-Learn, NLTK, [Keras](https://keras.io/), [TensorFlow](https://www.tensorflow.org/) and [PyTorch](). A [Vagrant](https://www.vagrantup.com/) file is used to generate this VM based on Ubuntu 16.04 (`xenial64`).

### Getting Started

*   This VM should work in Windows, macOS and Linux (tested in Debian-10 Buster)

*   VirtualBox (version 6+) is required

*   [Download and install Vagrant](http://www.vagrantup.com/downloads.html) 

*   Clone the virtual machine specs with:

```
git clone \
	https://github.com/f0nzie/vagrant-deeplearning-XE64G20U2R5120JPY383AE2007P8888.git
```

*   To generate the virtual machine, change to the folder where this repo has been cloned and type on your local terminal:

    ```
    vagrant up
    ```

* Browse to Jupyter with:  http://127.0.0.1:8888/. Try the different notebooks in there.

*   To access the virtual machine console or terminal, type:

    ```
    vagrant ssh
    ```


*   When finished, power off the virtual machine with:

    ```
    vagrant halt
    ```

    

### What's Installed

Anaconda3 Python 3.8.3

- Deep Learning
    - Keras 2.4.3
    - TensorFlow 2.2.0
    - PyTorch 1.7.1
- Machine Learning
    - sklearn 0.23.1
    - nltk  3.5
- Numeric and Data Science
    - numpy 1.18.5
    - scipy 1.4.1
    - matplotlib 3.2.2
    - seaborn 0.10.1
    - pandas 1.0.5
    - bokeh 2.1.1
    - h5py 2.10.0
    - xlsxwriter 1.2.9
    - statsmodels 0.11.1
    - regex 2020.6.8



## Jupyter notebook server

*   Jupyter notebook server is available at the host's browser at http://localhost:8888. 
*   Jupyter auto starts at boot by means of `systemd` `jupyter.service`
*   Service configuration file at `/etc/systemd/system/jupyter.service` generated by `provision-vagrant.sh` script.
*   Added PATH `/home/vagrant/.pyenv/versions/anaconda3-2020.07/bin` to `Environment` to be able to execute `conda` commands from Jupyter.



## Testing

There are several notebooks to test the deep learning packages:

*   [test_keras.ipynb](http://127.0.0.1:8100/notebooks/test_keras.ipynb)
*   [test_tensorflow.ipynb](http://127.0.0.1:8100/notebooks/test_tensorflow.ipynb)
*   [theano_test.ipynb](http://127.0.0.1:8100/notebooks/theano_test.ipynb)
*   [test_loop_packages.ipynb](http://127.0.0.1:8100/notebooks/test_loop_packages.ipynb): check the versions of the main installed packages


## Other

* Disk size: 20 GB
* RAM: 5120 GB
* CPUs: 2
* Network: NAT
* USB: off
* Shared folders: 0


## Codes machine name

* `XE64`: Ubuntu Xenial 64-bit
* `G20`: disk size 20 GB
* `U2`: 02 CPUs
* `R5120`: RAM 5120 MB
* `JPY383`: Jupyter server with Python 3.8.3
* `AE2007`: Anaconda3  version anaconda3-2020.07
* `P8888`: Jupyter port 8888



* Host name: `XE64G20U2R5120JPY383AE2007P8888`
* VM name: `vagrant-XE64G20U2R5120JPY383AE2007P8888`
* Short name: `XE64G20JPY383P8888`



## Credits

*   Thanks to [antonpirker](https://github.com/antonpirker/jupyter-vagrant-box) for the original Vagrant box.

*    [Florian Demmer](https://github.com/fdemmer) for the Vagrant provisioning.

## Notes

*   Using Anaconda3-2020.07 installed with `pyenv`
*   Python version 3.8.3



NEWS.md
# NEWS

# 20210116

*   Installed tensorflow 2.2, keras and pytorch:

    ```
    pip install tensorflow==2.2 keras    #! install tensorflow
    pip install torch
    ```

    All running:
    ![image-20210116182417413](assets/NEWS/image-20210116182417413.png)

*   Install jupyterhub in conda environment `anaconda3-2020.07`. Runs if installed manually. Fails from script built. Added port 8000 for jupyterhub:

    ```
    config.vm.network "forwarded_port", guest: 8000, host: 8000
    ```

    Linux dependencies for jupyterhub:

    ```
    # for JupyterHub
    apt-get install -y npm nodejs-legacy
    ```

    

*   Install Jupyter in a new pyenv environment in Python 3.7.9

    ```
    pyenv install 3.7.9
    pyenv virtualenv 3.7.9 deeplearning
    mkdir pytorch
    cd pytorch
    pyenv local deeplearning
    pip install tensorflow==2.2 keras    #! install tensorflow
    echo "install jupyter"
    pip install jupyter
    ```

    And in `provision-vagrant.sh`:

    ```
    ExecStart=/home/vagrant/.pyenv/versions/deeplearning/bin/jupyter notebook --port=8890 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/
    ```

    In Vagrantfile:

    ```
    config.vm.network "forwarded_port", guest: 8890, host: 8890
    ```

    Installs but it doesn't show in a browser.

*   Install tensorflow on pyenv environment with Python 3.7.9. Success!

    ```
    pip install tensorflow==2.2 keras
    ```

    ![image-20210116134155879](assets/NEWS/image-20210116134155879.png)

*   Add `deeplearning` environment with pyenv. Install `tensorflow==2.2`

    ```
    cd $HOME
    pyenv install 3.7.9
    pyenv virtualenv 3.7.9 deeplearning
    mkdir pytorch
    cd pytorch
    pyenv local deeplearning
    pip install tensorflow==2.2 keras
    ```

    

*   Add environment variable to `jupyter.service` file:

    ```
    Environment="PATH=/home/vagrant/.pyenv/versions/anaconda3-2020.07/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin"
    ```

    And execute jupyter with:

    ```
    ExecStart=/home/vagrant/.pyenv/versions/anaconda3-2020.07/bin/jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/data
    ```

    Now Jupyter runs on `anaconda3-2020.07`

*   TensorFlow does not install in Python 3,8.7 or Python 3.9.1

*   PyTorch install successfully with increased disk size under Python 3.7.9

*   Increasing disk size from 10 GB to 20 GB. Success!
    ![image-20210116122808608](assets/NEWS/image-20210116122808608.png)

*   Increasing RAM of Vagrant VM. FAILED!
    ![image-20210116120254993](assets/NEWS/image-20210116120254993.png)

*   Testing pyenv creating an isolated environment for `pytorch`:
    

*   Add Python 3.7.9:
    ![image-20210116100125598](assets/NEWS/image-20210116100125598.png)

*   Add `pyenv` dependencies in system provisioner:

    ```
    sudo apt-get install -y make build-essential libssl-dev zlib1g-dev \
    libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev \
    libncursesw5-dev xz-utils tk-dev libffi-dev liblzma-dev python-openssl
    ```

    

*   Going back to  `anaconda3-2019.03`. Working. **We leave it as is.**
    ![image-20210116092525302](assets/NEWS/image-20210116092525302.png)

![image-20210116092646811](assets/NEWS/image-20210116092646811.png)

* Trying pyenv `anaconda3-2019.10`. FAILED!

    ![image-20210116091330272](assets/NEWS/image-20210116091330272.png)

    ![image-20210116091352179](assets/NEWS/image-20210116091352179.png)

* Trying pyenv `anaconda3-2019.03`. **Success!**.
    ![image-20210116090430816](assets/NEWS/image-20210116090430816.png)
    
* Trying now with `anaconda3-2018.12`, the next version after `anaconda3-5.3.1`. Sucsess!
    ![image-20210116084858911](assets/NEWS/image-20210116084858911.png)
    
* Tried with pyenv `anaconda3-5.3.1`. **Success!** `conda` commands can run from Jupyter. This is the output of `!env`:
    ```
    USER=vagrant
    MKL_THREADING_LAYER=INTEL
    MPLBACKEND=module://ipykernel.pylab.backend_inline
    HOME=/home/vagrant
    PAGER=cat
    LC_CTYPE=C.UTF-8
    LOGNAME=vagrant
    TERM=xterm-color
    PATH=/home/vagrant/.pyenv/versions/anaconda3-5.3.1/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
    SHELL=/bin/bash
    GIT_PAGER=cat
    CLICOLOR=1
    PWD=/vagrant
    JPY_PARENT_PID=17184
    ```

# 20210115
* .`bash_profile` generated by the provisioning script:
    ```
    export PATH="/home/vagrant/.pyenv/bin:$PATH"
    eval "$(pyenv init -)"
    eval "$(pyenv virtualenv-init -)"
    ```
    This line:
    ```
    grep -q -F "export PATH=\"/home/vagrant/.pyenv/bin:\$PATH\"" ~/.bash_profile || echo "export PATH=\"/home/vagrant/.pyenv/bin:\$PATH\"" >> ~/.bash_profile
    ```
    generates `export PATH="/home/vagrant/.pyenv/bin:$PATH"`

    This other line:
    ```
    grep -q -F "eval \"\$(pyenv init -)\"" ~/.bash_profile || echo "eval \"\$(pyenv init -)\"" >> ~/.bash_profile
    ```
    generates this `eval "$(pyenv init -)"`

    And this other line:
    ```
    grep -q -F "eval \"\$(pyenv virtualenv-init -)\"" ~/.bash_profile || echo "eval \"\$(pyenv virtualenv-init -)\"" >> ~/.bash_profile
    ```
    generates this: `eval "$(pyenv virtualenv-init -)"`


* This is the result of running !env in a notebook under `anaconda3-5.1.0`:
    ```
    USER=vagrant
    MKL_THREADING_LAYER=INTEL
    MPLBACKEND=module://ipykernel.pylab.backend_inline
    HOME=/home/vagrant
    PAGER=cat
    LOGNAME=vagrant
    TERM=xterm-color
    PATH=/home/vagrant/.pyenv/versions/anaconda3-5.1.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
    SHELL=/bin/bash
    GIT_PAGER=cat
    CLICOLOR=1
    PWD=/vagrant
    JPY_PARENT_PID=13474
    ```
* With  `anaconda3-2020.07` shell calls to conda from Jupyter notebooks will not work.
    ![image-20210115203701689](assets/NEWS/image-20210115203701689.png)
* Switch Anaconda from `anaconda3-5.1.0` to `anaconda3-2020.07`

scripts/provision-vagrant.sh
#!/bin/bash

echo "Install system requirements"
apt-get --quiet update
apt-get install -y --no-install-recommends \
	curl \
	git

# for pyenv
apt-get install -y make build-essential libssl-dev zlib1g-dev \
libbz2-dev libreadline-dev libsqlite3-dev wget llvm libncurses5-dev \
libncursesw5-dev xz-utils tk-dev libffi-dev liblzma-dev python-openssl

# for JupyterHub
apt-get install -y npm nodejs-legacy

echo "Install pyenv"
curl -# -L https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash

echo "Setup home directory"
ln -s /vagrant/ /home/vagrant/data        # symbolic link to host files on folder data
chown vagrant.vagrant /home/vagrant/data

echo "Setup Jupyter auto start"
cat >/etc/systemd/system/jupyter.service <<EOL
[Unit]
Description=Jupyter Workplace

[Service]
Type=simple
PIDFile=/run/jupyter.pid
Environment="PATH=/home/vagrant/.pyenv/versions/anaconda3-2020.07/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin"
# ExecStart=/home/vagrant/.pyenv/versions/anaconda3-5.1.0/bin/jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/data
# ExecStart=/home/vagrant/.pyenv/versions/anaconda3-5.3.1/bin/jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/data
# ExecStart=/home/vagrant/.pyenv/versions/anaconda3-2018.12/bin/jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/data
# ExecStart=/home/vagrant/.pyenv/versions/anaconda3-2019.03/bin/jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/data
# ExecStart=/home/vagrant/.pyenv/versions/anaconda3-2019.10/bin/jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/data
ExecStart=/home/vagrant/.pyenv/versions/anaconda3-2020.07/bin/jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/data
# ExecStart=/home/vagrant/.pyenv/versions/deeplearning/bin/jupyter notebook --port=8890 --no-browser --ip=0.0.0.0 --NotebookApp.token= --notebook-dir=/home/vagrant/
User=vagrant
Group=vagrant
WorkingDirectory=/home/vagrant/data
Restart=always
RestartSec=10

[Install]
WantedBy=multi-user.target
EOL

# export PATH="/home/vagrant/.pyenv/versions/anaconda3-2020.07/bin:$PATH"

systemctl enable jupyter.service
systemctl daemon-reload
systemctl restart jupyter.service


scripts/provision-vagrant-user.sh
#!/bin/bash

echo "Install pyenv"
touch ~/.bash_profile
grep -q -F "export PATH=\"/home/vagrant/.pyenv/bin:\$PATH\"" ~/.bash_profile || echo "export PATH=\"/home/vagrant/.pyenv/bin:\$PATH\"" >> ~/.bash_profile
source ~/.bash_profile
grep -q -F "eval \"\$(pyenv init -)\"" ~/.bash_profile || echo "eval \"\$(pyenv init -)\"" >> ~/.bash_profile
grep -q -F "eval \"\$(pyenv virtualenv-init -)\"" ~/.bash_profile || echo "eval \"\$(pyenv virtualenv-init -)\"" >> ~/.bash_profile
curl -s -S -L https://raw.githubusercontent.com/yyuu/pyenv-installer/master/bin/pyenv-installer | bash
source ~/.bash_profile

echo "Install Anaconda3 with pyenv)"
# pyenv install anaconda3-5.1.0      #? Python 3.6.4
# pyenv install anaconda3-5.3.1      #? Python 
pyenv install anaconda3-2020.07      #? conda from July 2020. Python 3.8.3
# pyenv install anaconda3-2018.12  #! Success
# pyenv install anaconda3-2019.03    #! Success
# pyenv install anaconda3-2019.10    #! FAILED 


echo "Activate anaconda as local pyenv"   
cd /home/vagrant/data
# pyenv local anaconda3-5.1.0     #? set folder with specified environment Python 3.6.4
# pyenv local anaconda3-5.3.1     #? set folder with specified environment Python 3.6.4
pyenv local anaconda3-2020.07   #? set folder with specified environment Python 3.6.4. FAILED
# pyenv local anaconda3-2018.12   #! Success
# pyenv local anaconda3-2019.03     #! Success
# pyenv local anaconda3-2019.10    #! FAILED

pip install tensorflow==2.2 keras    #! install tensorflow
pip install torch

# echo "create deeplearning environment with pyenv 3.7.9"
# cd $HOME
# pyenv install 3.7.9
# pyenv virtualenv 3.7.9 deeplearning
# mkdir pytorch
# cd pytorch
# pyenv local deeplearning
# # pip install tensorflow==2.2 keras    #! install tensorflow

# echo "jupyterhub"
# conda install -c conda-forge jupyterhub
# conda install notebook
# conda install jupyterlab
# sudo mkdir /etc/jupyterhub
# sudo /home/vagrant/.pyenv/versions/deeplearning/bin/jupyterhub --generate-config -f /etc/jupyterhub/jupyterhub_config.py
# sudo ~/.pyenv/versions/deeplearning/bin/jupyterhub &
# # echo "install jupyter"
# # pip install jupyter
# # jupyter-notebook --port 8889 --no-browser &    #! doesn't work
# cd $HOME

echo "Setup login directory"
echo "cd /home/vagrant/data/"  >> /home/vagrant/.bash_profile
echo "source ~/.bashrc"  >> /home/vagrant/.bash_profile
